arch/sim/src/sim/posix/sim_host_v4l2.c

```diff
@@ -0,0 +1,349 @@
+  host_video_dev_t *dev = (host_video_dev_t *)vdev;
+
+  memset(&buf, 0, sizeof(buf));
+  buf.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+  buf.memory = V4L2_MEMORY_MMAP;
+
+  /* Dequeue a buffer */
+
+  if (-1 == host_video_ioctl(dev->fd, VIDIOC_DQBUF, &buf))
+    {
+      switch (errno)
+        {
+          case EAGAIN:
+
+            /* No buffer in the outgoing queue */
+
+            return 0;
+          case EIO:
+
+            /* fall through */
+
+          default:
+            perror("VIDIOC_DQBUF");
+            return -errno;
+        }
+    }
+
+  if (size > buf.bytesused)
+    {
+      size = buf.bytesused;
+    }
+
+  memcpy(addr, dev->addrs[buf.index], size);
+  if (-1 == ioctl(dev->fd, VIDIOC_QBUF, &buf))
+    {
+      perror("VIDIOC_QBUF");
+      return -errno;
+    }
+
+  return size;
+}
+
+int host_video_uninit(host_video_dev vdev)
+{
+  host_video_dev_t *dev = (host_video_dev_t *)vdev;
+
+  if (dev != NULL)
+    {
+      close(dev->fd);
+    }
+
+  free(vdev);
+  return 0;
+}
+
+int host_video_start_capture(host_video_dev vdev)
+{
+  struct v4l2_buffer buf;
+  enum v4l2_buf_type type;
+  int i;
+  host_video_dev_t *dev = (host_video_dev_t *)vdev;
+
+  /* VIDIOC_REQBUFS initiate user pointer I/O */
+
+  dev->reqbuf.type   = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+  dev->reqbuf.memory = V4L2_MEMORY_MMAP;
+  dev->reqbuf.count  = MAX_REQBUFS;
+
+  if (-1 == host_video_ioctl(dev->fd, VIDIOC_REQBUFS, &dev->reqbuf))
+    {
+      perror("VIDIOC_REQBUFS");
+      return -errno;
+    }
+
+  if (dev->reqbuf.count < 2)
+    {
+      errno = ENOMEM;
+      perror("Not enough buffers");
+      return -ENOMEM;
+    }
+
+  for (i = 0; i < dev->reqbuf.count; i++)
+    {
+      memset(&buf, 0, sizeof(buf));
+      buf.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+      buf.memory = V4L2_MEMORY_MMAP;
+      buf.index = i;
+      if (-1 == host_video_ioctl(dev->fd, VIDIOC_QUERYBUF, &buf))
+        {
+          perror("VIDIOC_QUERYBUF");
+          return -errno;
+        }
+
+      dev->addrs[i] = mmap(NULL, buf.length, PROT_READ | PROT_WRITE,
+                           MAP_SHARED, dev->fd, buf.m.offset);
+      dev->buflen[i] = buf.length;
+      if (dev->addrs[i] == MAP_FAILED)
+        {
+          perror("Mmap failed");
+          return -errno;
+        }
+
+      if (-1 == host_video_ioctl(dev->fd, VIDIOC_QBUF, &buf))
+        {
+          perror("VIDIOC_QBUF");
+          return -errno;
+        }
+    }
+
+  type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+  if (-1 == host_video_ioctl(dev->fd, VIDIOC_STREAMON, &type))
+    {
+      perror("VIDIOC_STREAMON");
+      return -errno;
+    }
+
+  return 0;
+}
+
+int host_video_stop_capture(host_video_dev vdev)
+{
+  enum v4l2_buf_type type;
+  int i;
+  host_video_dev_t *dev = (host_video_dev_t *)vdev;
+
+  type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+  if (-1 == host_video_ioctl(dev->fd, VIDIOC_STREAMOFF, &type))
+    {
+      perror("VIDIOC_STREAMOFF");
+      return -errno;
+    }
+
+  for (i = 0; i < dev->reqbuf.count; i++)
```

Comment:
change reqbuf to local variable in host_video_start_capture
```suggestion
  for (i = 0; i < MAX_REQBUFS;  i++)
    {
      if (dev->addrs[i], != NULL && dev->addr[i] != MAP_FAILED)
    }
```